R/paste.R

```diff
@@ -3,78 +3,214 @@
+
+  # combine dm and paste code
+  code_dm <- glue_collapse(
+    c(
+      code_construct,
+      code_select,
+      code_pks,
+      code_fks,
+      code_color
+    ),
+    sep = glue(" %>%\n{tab}", .trim = FALSE)
+  )
+
+  paste0(code_tables, code_dm)
+}
+
+dm_paste_tables <- function(dm, tab) {
+  ptype <- dm_ptype(dm)
+
+  tables <-
+    ptype %>%
+    dm_get_tables() %>%
+    map_chr(df_paste, tab)
+
+  glue_collapse1(
+    glue("{tick_if_needed(names(tables))} <- {tables}\n\n", .trim = FALSE)
+  )
+}
+
+dm_paste_construct <- function(dm) {
+  glue("dm({glue_collapse1(tick_if_needed(src_tbls(dm)), ', ')})")
+}
+
+dm_paste_select <- function(dm) {
+  tbl_select <- dm %>%
+    dm_get_def() %>%
+    mutate(cols = map(data, colnames)) %>%
+    mutate(cols = map_chr(cols, ~ glue_collapse1(glue(", {tick_if_needed(.x)}")))) %>%
+    mutate(code = glue("dm_select({tick_if_needed(table)}{cols})")) %>%
+    pull()
+}
+
+dm_paste_pks <- function(dm) {
+  # FIXME: this will fail with compound keys
+  dm_get_all_pks_impl(dm) %>%
+    mutate(code = glue("dm_add_pk({tick_if_needed(table)}, {tick_if_needed(pk_col)})")) %>%
+    pull()
+}
+
+dm_paste_fks <- function(dm) {
   # FIXME: this will fail with compound keys
   tbl_fks <- dm_get_all_fks_impl(dm) %>%
-    mutate(code = glue("{tab}dm_add_fk({child_table}, {child_fk_cols}, {parent_table})"))
-  code_fks <- if (nrow(tbl_fks)) summarize(tbl_fks, code = glue_collapse(code, sep = " %>%\n")) %>% pull() else character()
+    mutate(code = glue("dm_add_fk({tick_if_needed(child_table)}, {tick_if_needed(child_fk_cols)}, {tick_if_needed(parent_table)})")) %>%
+    pull()
+}
 
-  if (color) {
-    colors <- dm_get_colors(dm)
-    colors <- colors[names(colors) != "default"]
-    code_color <- imap_chr(colors, ~ glue("{tab}dm_set_colors({tick_if_needed(..2)} = {tick_if_needed(..1)})"))
+dm_paste_color <- function(dm) {
+  colors <- dm_get_colors(dm)
+  colors <- colors[names(colors) != "default"]
+  glue("dm_set_colors({tick_if_needed(names(colors))} = {tick_if_needed(colors)})")
+}
+
+df_paste <- function(x, tab) {
+  cols <- map_chr(x, deparse_line)
+  if (is_empty(x)) {
+    cols <- ""
+  } else {
+    cols <- paste0(
+      paste0("\n", tab, tick_if_needed(names(cols)), " = ", cols, collapse = ","),
+      "\n"
+    )
+  }
+
+  paste0("tibble(", cols, ")")
+}
+
+deparse_line <- function(x) {
+  x <- deparse(x, width.cutoff = 500, backtick = TRUE)
+  gsub(" *\n *", " ", x)
+}
+
+glue_collapse1 <- function(x, ...) {
+  if (is_empty(x)) {
+    ""
   } else {
-    code_color <- character()
+    glue_collapse(x, ...)
+  }
+}
+
+dquote <- function(x) {
+  if (is_empty(x)) {
+    return(character())
   }
+  paste0('"', x, '"')
+}
+
+# Errors ------------------------------------------------------------------
+
+abort_unknown_option <- function(options, all_options) {
+  abort(error_txt_unknown_option(options, all_options), .subclass = dm_error_full("unknown_option"))
+}
```

Comment:
Move this and `error_txt_unknown_option()` to error-helpers.R?

Reply:
No, `error-helpers.R` should disappear at some point, all error constructors should be moved to the files where they are used.